csharp/src/Ice/Communicator-OutgoingConnectionFactory.cs

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) ZeroC, Inc. All rights reserved.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace ZeroC.Ice
+{
+    public sealed partial class Communicator : IConnectionManager
+    {
+        private readonly MultiDictionary<(Endpoint, NonSecure, object), Connection> _activeConnectionMap = new(ConnectionMapKeyComparer.Instance);
+        private readonly Dictionary<(Endpoint, NonSecure, object), Task<Connection>> _pendingConnectionMap = new(ConnectionMapKeyComparer.Instance);
+        // We keep a map of the endpoints that recently resulted in a failure while establishing a connection. This is
+        // used to influence the selection of endpoints when creating new connections. Endpoints with recent failures
+        // are tried last.
+        private readonly ConcurrentDictionary<Endpoint, DateTime> _transportFailures = new();
+
+        public void Remove(Connection connection)
+        {
+            lock (_mutex)
+            {
+                _activeConnectionMap.Remove((connection.Endpoint, connection.PreferNonSecure, connection.ConnectionId),
+                                            connection);
+            }
+        }
+
+        internal async ValueTask<Connection> ConnectAsync(
+            Endpoint endpoint,
+            NonSecure preferNonSecure,
+            object cookie,
+            CancellationToken cancel)
+        {
+            Task<Connection>? connectTask;
+            lock (_mutex)
+            {
+                // Check if there is an active connection that we can use.
+                if (_activeConnectionMap.TryGetValue((endpoint, preferNonSecure, cookie),
+                                                      out ICollection<Connection>? connections) &&
+                    connections.FirstOrDefault(connection => connection.IsActive) is Connection connection)
+                {
```

Comment:
Yes. However, I think we should also check for NonSecure=Never, i.e.
a) first lookup exact match for preferNonSecure value
b) then when preferNonSecure != Never, also lookup for Never

Reply:
Maybe is simpler to keep IsSecure bool as part of the key for estalbish connections, and just check if the connection is valid for the preferences. For example here if the remote endpoint only accepts secure connections but 3 enpoints has Always, SameHost, and TrustedHost you still end-up with 3 secure connections when one should be enough.

Reply:
I was thinking that you would create 2 entries in the main connection map in this case (preferNonSecure != Never results in a secure connection):
- one with the requested preferNonSecure value
- another one with preferNonSecure=Never

And naturally use the same logic when removing from the map.

When the resulting connection is non-secure, we could likewise insert an entry with preferNonSecure=Always. This way, each active connection has an entry with preferNonSecure=Never or preferNonSecure=Always, which makes it easier for subsequent lookups.

Reply:
The downside of this idea is (a) it's complicated and (b) we have to keep PreferNonSecure on the connection to eventually remove the connection from the dictionary. Maybe it would be better to switch to dictionary like:
```
private readonly Dictionary<(Endpoint, object), List<(Connection, NonSecure)>> _activeConnections = ...
```
This way, we don't need to keep PreferNonSecure on the connection: to remove a connection, we just iterate over the list (shouldn't be very long). Likewise, we would simply iterate over this short list when looking up an existing connection.